src/app/views/query-runner/query-input/AutoComplete.tsx

```diff
@@ -0,0 +1,264 @@
+import { getTheme, KeyCodes, Label, TextField } from 'office-ui-fabric-react';
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import { bindActionCreators, Dispatch } from 'redux';
+import { IAutoCompleteProps, IAutoCompleteState } from '../../../../types/auto-complete';
+import * as autoCompleteActionCreators from '../../../services/actions/autocomplete-action-creators';
+import { parseSampleUrl } from '../../../utils/sample-url-generation';
+import { queryInputStyles } from './QueryInput.styles';
+import { cleanUpSelectedSuggestion } from './util';
+
+class AutoComplete extends Component<IAutoCompleteProps, IAutoCompleteState> {
+
+  constructor(props: IAutoCompleteProps) {
+    super(props);
+
+    this.state = {
+      activeSuggestion: 0,
+      filteredSuggestions: [],
+      suggestions: [],
+      showSuggestions: false,
+      userInput: this.props.sampleQuery.sampleUrl,
+      compare: ''
+    };
+  }
+
+  public onChange = (e: any) => {
+    const { suggestions, showSuggestions, userInput: previousUserInput, compare } = this.state;
+    const userInput = e.target.value;
+
+    this.props.contentChanged(userInput);
+
+    this.setState({
+      userInput
+    });
+
+    if (showSuggestions && suggestions.length) {
+      this.filterSuggestions(userInput, previousUserInput, compare, suggestions);
+    }
+    this.initialiseAutoComplete(userInput);
+  };
+
+  public onClick = (e: any) => {
+    const selected = e.currentTarget.innerText;
+    this.appendSuggestionToUrl(selected);
+  };
+
+  private initialiseAutoComplete = (url: string) => {
+    const lastCharacter = url.substring(url.length - 1);
+    switch (lastCharacter) {
+      case '?':
+        this.requestForAutocompleteOptions(url);
+        break;
+
+      case '=':
+        this.getParameterEnums(url);
+        break;
+
+      case ',':
+        this.getParameterEnums(url);
+        break;
+
+      case '&':
+        this.getPathParameters();
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  public onKeyDown = (e: any) => {
+    const { activeSuggestion, filteredSuggestions, showSuggestions } = this.state;
+
+    switch (e.keyCode) {
+      case KeyCodes.tab:
```

Comment:
Pressing tab appends the suggestion to the url then transfers focus to the `Run Query` button. Can we maintain the focus in QueryInput?

Reply:
Yes. I plan on implementing it.